File: src/main/java/com/skysoft/gui/scale/InventoryCursorMemory.java
package com.skysoft.gui.scale;

import com.mojang.blaze3d.platform.InputConstants;
import com.mojang.blaze3d.platform.Window;
import com.skysoft.config.SkysoftConfigGui;
import net.minecraft.client.gui.screens.Screen;
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen;

public final class InventoryCursorMemory {
    private static final long EXPIRY_NANOS = 1_000_000_000L;
    private static final double CENTER_TOLERANCE = 1.0;
    private static final int RESTORE_ATTEMPTS = 3;

    private static CursorPoint cursorBeforeGrab;
    private static CursorSnapshot snapshot;
    private static int remainingRestores;

    private InventoryCursorMemory() {
    }

    public static void rememberScreenCursor(Screen screen, double x, double y) {
        if (!canRemember(screen)) {
            discard();
            return;
        }
        snapshot = new CursorSnapshot(new CursorPoint(x, y), null, System.nanoTime());
        remainingRestores = 0;
    }

    public static void beginMouseGrab(double x, double y) {
        if (!isEnabled()) {
            discard();
            return;
        }
        cursorBeforeGrab = new CursorPoint(x, y);
    }

    public static void finishMouseGrab(double centerX, double centerY) {
        if (!isEnabled()) {
            discard();
            return;
        }
        if (cursorBeforeGrab == null) {
            return;
        }
        snapshot = new CursorSnapshot(cursorBeforeGrab, new CursorPoint(centerX, centerY), System.nanoTime());
        cursorBeforeGrab = null;
    }

    public static CursorPoint cursorForRelease(double centerX, double centerY) {
        if (!isUsable() || snapshot.grabCenter() == null) {
            return null;
        }
        CursorPoint expected = snapshot.grabCenter();
        if (Math.abs(expected.x() - centerX) >= CENTER_TOLERANCE ||
            Math.abs(expected.y() - centerY) >= CENTER_TOLERANCE) {
            discard();
            return null;
        }

        remainingRestores = RESTORE_ATTEMPTS;
        snapshot = new CursorSnapshot(snapshot.cursor(), null, snapshot.capturedAt());
        return snapshot.cursor();
    }

    public static void restoreWhenScreenInitializes(Screen screen, Window window, CursorController cursor) {
        if (!canRestore(screen)) {
            return;
        }
        remainingRestores = RESTORE_ATTEMPTS;
        restoreOnce(window, cursor);
    }

    public static void continueRestore(Screen screen, Window window, CursorController cursor) {
        if (remainingRestores <= 0) {
            return;
        }
        if (!canRestore(screen)) {
            return;
        }
        restoreOnce(window, cursor);
    }

    private static void restoreOnce(Window window, CursorController cursor) {
        CursorPoint point = snapshot.cursor();
        cursor.skysoft$moveCursor(point.x(), point.y());
        InputConstants.grabOrReleaseMouse(window, InputConstants.CURSOR_NORMAL, point.x(), point.y());
        remainingRestores--;
        if (remainingRestores == 0) {
            discard();
        }
    }

    private static boolean canRemember(Screen screen) {
        return isEnabled() && screen instanceof AbstractContainerScreen<?>;
    }

    private static boolean canRestore(Screen screen) {
        if (!canRemember(screen)) {
            discard();
            return false;
        }
        if (!isUsable()) {
            discard();
            return false;
        }
        return true;
    }

    private static boolean isUsable() {
        return snapshot != null && System.nanoTime() - snapshot.capturedAt() <= EXPIRY_NANOS;
    }

    private static boolean isEnabled() {
        return SkysoftConfigGui.INSTANCE.config().inventory.preserveCursorPosition;
    }

    private static void discard() {
        cursorBeforeGrab = null;
        snapshot = null;
        remainingRestores = 0;
    }

    private record CursorSnapshot(CursorPoint cursor, CursorPoint grabCenter, long capturedAt) {
    }

    public record CursorPoint(double x, double y) {
    }
}

File: src/main/kotlin/com/skysoft/gui/scale/InventoryCursorMemory.kt
package com.skysoft.gui.scale

import com.mojang.blaze3d.platform.InputConstants
import com.mojang.blaze3d.platform.Window
import com.skysoft.config.SkysoftConfigGui
import net.minecraft.client.gui.screens.Screen
import net.minecraft.client.gui.screens.inventory.AbstractContainerScreen

class InventoryCursorMemory private constructor() {
    @JvmRecord
    data class CursorPoint(val x: Double, val y: Double)

    private data class CursorSnapshot(
        val cursor: CursorPoint,
        val grabCenter: CursorPoint?,
        val capturedAt: Long,
    )

    companion object {
        private const val EXPIRY_NANOS = 1_000_000_000L
        private const val CENTER_TOLERANCE = 1.0
        private const val RESTORE_ATTEMPTS = 3

        private var cursorBeforeGrab: CursorPoint? = null
        private var snapshot: CursorSnapshot? = null
        private var remainingRestores = 0

        @JvmStatic
        fun rememberScreenCursor(screen: Screen?, x: Double, y: Double) {
            if (!canRemember(screen)) {
                discard()
                return
            }
            snapshot = CursorSnapshot(CursorPoint(x, y), null, System.nanoTime())
            remainingRestores = 0
        }

        @JvmStatic
        fun beginMouseGrab(x: Double, y: Double) {
            if (!isEnabled()) {
                discard()
                return
            }
            cursorBeforeGrab = CursorPoint(x, y)
        }

        @JvmStatic
        fun finishMouseGrab(centerX: Double, centerY: Double) {
            if (!isEnabled()) {
                discard()
                return
            }
            val cursor = cursorBeforeGrab ?: return
            snapshot = CursorSnapshot(cursor, CursorPoint(centerX, centerY), System.nanoTime())
            cursorBeforeGrab = null
        }

        @JvmStatic
        fun cursorForRelease(centerX: Double, centerY: Double): CursorPoint? {
            if (!isUsable()) return null
            val current = checkNotNull(snapshot)
            val expected = current.grabCenter ?: return null
            if (
                kotlin.math.abs(expected.x - centerX) >= CENTER_TOLERANCE ||
                kotlin.math.abs(expected.y - centerY) >= CENTER_TOLERANCE
            ) {
                discard()
                return null
            }
            remainingRestores = RESTORE_ATTEMPTS
            snapshot = CursorSnapshot(current.cursor, null, current.capturedAt)
            return current.cursor
        }

        @JvmStatic
        fun restoreWhenScreenInitializes(screen: Screen?, window: Window, cursor: CursorController) {
            if (!canRestore(screen)) return
            remainingRestores = RESTORE_ATTEMPTS
            restoreOnce(window, cursor)
        }

        @JvmStatic
        fun continueRestore(screen: Screen?, window: Window, cursor: CursorController) {
            if (remainingRestores <= 0 || !canRestore(screen)) return
            restoreOnce(window, cursor)
        }

        private fun restoreOnce(window: Window, cursor: CursorController) {
            val point = checkNotNull(snapshot).cursor
            cursor.skysoftMoveCursor(point.x, point.y)
            InputConstants.grabOrReleaseMouse(window, InputConstants.CURSOR_NORMAL, point.x, point.y)
            remainingRestores--
            if (remainingRestores == 0) discard()
        }

        private fun canRemember(screen: Screen?): Boolean =
            isEnabled() && screen is AbstractContainerScreen<*>

        private fun canRestore(screen: Screen?): Boolean {
            if (!canRemember(screen) || !isUsable()) {
                discard()
                return false
            }
            return true
        }

        private fun isUsable(): Boolean =
            snapshot?.let { System.nanoTime() - it.capturedAt <= EXPIRY_NANOS } == true

        private fun isEnabled(): Boolean = SkysoftConfigGui.config().inventory.preserveCursorPosition

        private fun discard() {
            cursorBeforeGrab = null
            snapshot = null
            remainingRestores = 0
        }
    }
}
